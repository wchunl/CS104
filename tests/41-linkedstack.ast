ROOT "" (0.0.0)
|   TOK_FUNCTION "" (6.22.1) function
|   |   TOK_TYPE_ID "" (6.22.1) typeid
|   |   |   TOK_VOID "void" (6.22.1) void
|   |   |   TOK_IDENT "fail" (6.22.6) variable
|   |   TOK_PARAM "(" (6.22.11) param
|   |   |   TOK_TYPE_ID "" (6.22.12) typeid
|   |   |   |   TOK_STRING "string" (6.22.12) string
|   |   |   |   TOK_IDENT "expr" (6.22.19) string variable (6.22.12)
|   |   |   TOK_TYPE_ID "" (6.22.25) typeid
|   |   |   |   TOK_STRING "string" (6.22.25) string
|   |   |   |   TOK_IDENT "file" (6.22.32) string variable (6.22.25)
|   |   |   TOK_TYPE_ID "" (6.22.38) typeid
|   |   |   |   TOK_INT "int" (6.22.38) int
|   |   |   |   TOK_IDENT "line" (6.22.42) int variable (6.22.38)
|   TOK_FUNCTION "" (6.24.1) function
|   |   TOK_TYPE_ID "" (6.24.1) typeid
|   |   |   TOK_VOID "void" (6.24.1) void
|   |   |   TOK_IDENT "putchr" (6.24.6) variable
|   |   TOK_PARAM "(" (6.24.13) param
|   |   |   TOK_TYPE_ID "" (6.24.14) typeid
|   |   |   |   TOK_INT "int" (6.24.14) int
|   |   |   |   TOK_IDENT "chr" (6.24.18) int variable (6.24.14)
|   TOK_FUNCTION "" (6.25.1) function
|   |   TOK_TYPE_ID "" (6.25.1) typeid
|   |   |   TOK_VOID "void" (6.25.1) void
|   |   |   TOK_IDENT "putint" (6.25.6) variable
|   |   TOK_PARAM "(" (6.25.13) param
|   |   |   TOK_TYPE_ID "" (6.25.14) typeid
|   |   |   |   TOK_INT "int" (6.25.14) int
|   |   |   |   TOK_IDENT "num" (6.25.18) int variable (6.25.14)
|   TOK_FUNCTION "" (6.26.1) function
|   |   TOK_TYPE_ID "" (6.26.1) typeid
|   |   |   TOK_VOID "void" (6.26.1) void
|   |   |   TOK_IDENT "putstr" (6.26.6) variable
|   |   TOK_PARAM "(" (6.26.13) param
|   |   |   TOK_TYPE_ID "" (6.26.14) typeid
|   |   |   |   TOK_STRING "string" (6.26.14) string
|   |   |   |   TOK_IDENT "str" (6.26.21) string variable (6.26.14)
|   TOK_FUNCTION "" (6.28.1) function
|   |   TOK_TYPE_ID "" (6.28.1) typeid
|   |   |   TOK_INT "int" (6.28.1) int
|   |   |   TOK_IDENT "getchr" (6.28.5) variable
|   |   TOK_PARAM "(" (6.28.11) param
|   TOK_FUNCTION "" (6.29.1) function
|   |   TOK_TYPE_ID "" (6.29.1) typeid
|   |   |   TOK_STRING "string" (6.29.1) string
|   |   |   TOK_IDENT "getstr" (6.29.8) variable
|   |   TOK_PARAM "(" (6.29.14) param
|   TOK_FUNCTION "" (6.30.1) function
|   |   TOK_TYPE_ID "" (6.30.1) typeid
|   |   |   TOK_STRING "string" (6.30.1) string
|   |   |   TOK_IDENT "getln" (6.30.8) variable
|   |   TOK_PARAM "(" (6.30.13) param
|   TOK_STRUCT "struct" (7.5.1)
|   |   TOK_IDENT "node" (7.5.8) variable
|   |   TOK_STRING "string" (7.6.4) string
|   |   |   TOK_TYPE_ID "data" (7.6.11) typeid
|   |   TOK_PTR "ptr" (7.7.4)
|   |   |   TOK_TYPE_ID "node" (7.7.16) typeid
|   |   |   TOK_TYPE_ID "link" (7.7.22) typeid
|   TOK_STRUCT "struct" (7.10.1)
|   |   TOK_IDENT "stack" (7.10.8) variable (7.14.12)
|   |   TOK_PTR "ptr" (7.11.4)
|   |   |   TOK_TYPE_ID "node" (7.11.16) typeid
|   |   |   TOK_TYPE_ID "top" (7.11.22) typeid
|   TOK_FUNCTION "" (7.14.1) function
|   |   TOK_TYPE_ID "" (7.14.1) typeid
|   |   |   TOK_INT "int" (7.14.1) int
|   |   |   TOK_IDENT "empty" (7.14.5) variable
|   |   TOK_PARAM "(" (7.14.11) param
|   |   |   TOK_TYPE_ID "" (7.14.12) typeid
|   |   |   |   TOK_PTR "ptr" (7.14.12)
|   |   |   |   |   TOK_TYPE_ID "stack" (7.14.23) typeid
|   |   |   |   TOK_IDENT "stack" (7.14.30) variable (7.14.12)
|   |   TOK_BLOCK "{" (7.14.37)
|   |   |   TOK_BLOCK "{" (7.15.4)
|   |   |   |   TOK_IF "if" (7.15.5)
|   |   |   |   |   TOK_NOT "not" (7.15.9)
|   |   |   |   |   |   TOK_NE "!=" (7.15.20)
|   |   |   |   |   |   |   TOK_IDENT "stack" (7.15.14) variable (7.14.12)
|   |   |   |   |   |   |   TOK_NULLPTR "nullptr" (7.15.23) null const
|   |   |   |   |   TOK_CALL "(" (7.15.38)
|   |   |   |   |   |   TOK_IDENT "fail" (7.15.33) variable
|   |   |   |   |   |   TOK_STRINGCON ""stack != nullptr"" (7.15.39) string const
|   |   |   |   |   |   |   TOK_STRINGCON ""41-linkedstack.oc"" (7.15.59) string const
|   |   |   |   |   |   |   TOK_INTCON "15" (7.15.80) int const
|   |   |   ';' ";" (7.15.85)
|   |   |   TOK_RETURN "return" (7.16.4)
|   |   |   |   TOK_EQ "==" (7.16.22) vreg
|   |   |   |   |   TOK_ARROW "->" (7.16.16) lval vaddr
|   |   |   |   |   |   TOK_IDENT "stack" (7.16.11) variable (7.14.12)
|   |   |   |   |   |   TOK_IDENT "top" (7.16.18) variable
|   |   |   |   |   TOK_NULLPTR "nullptr" (7.16.25) null const
|   TOK_FUNCTION "" (7.19.1) function
|   |   TOK_TYPE_ID "" (7.19.1) typeid
|   |   |   TOK_PTR "ptr" (7.19.1)
|   |   |   |   TOK_TYPE_ID "stack" (7.19.12) typeid
|   |   |   TOK_IDENT "new_stack" (7.19.19) variable
|   |   TOK_PARAM "(" (7.19.28) param
|   |   TOK_BLOCK "{" (7.19.31)
|   |   |   TOK_VARDECL "=" (7.20.28) variable lval
|   |   |   |   TOK_PTR "ptr" (7.20.4)
|   |   |   |   |   TOK_TYPE_ID "stack" (7.20.15) typeid
|   |   |   |   TOK_IDENT "stack" (7.20.22) variable (7.14.12)
|   |   |   |   TOK_ALLOC "alloc" (7.20.30)
|   |   |   |   |   TOK_TYPE_ID "stack" (7.20.43) typeid
|   |   |   '=' "=" (7.21.15) int vreg
|   |   |   |   TOK_ARROW "->" (7.21.9) lval vaddr
|   |   |   |   |   TOK_IDENT "stack" (7.21.4) variable (7.14.12)
|   |   |   |   |   TOK_IDENT "top" (7.21.11) variable
|   |   |   |   TOK_NULLPTR "nullptr" (7.21.17) null const
|   |   |   TOK_RETURN "return" (7.22.4)
|   |   |   |   TOK_IDENT "stack" (7.22.11) variable (7.14.12)
|   TOK_FUNCTION "" (7.25.1) function
|   |   TOK_TYPE_ID "" (7.25.1) typeid
|   |   |   TOK_VOID "void" (7.25.1) void
|   |   |   TOK_IDENT "push" (7.25.6) variable
|   |   TOK_PARAM "(" (7.25.11) param
|   |   |   TOK_TYPE_ID "" (7.25.12) typeid
|   |   |   |   TOK_PTR "ptr" (7.25.12)
|   |   |   |   |   TOK_TYPE_ID "stack" (7.25.23) typeid
|   |   |   |   TOK_IDENT "stack" (7.25.30) variable (7.14.12)
|   |   |   TOK_TYPE_ID "" (7.25.37) typeid
|   |   |   |   TOK_STRING "string" (7.25.37) string
|   |   |   |   TOK_IDENT "str" (7.25.44) string variable (6.26.14)
|   |   TOK_BLOCK "{" (7.25.49)
|   |   |   TOK_VARDECL "=" (7.26.26) variable lval
|   |   |   |   TOK_PTR "ptr" (7.26.4)
|   |   |   |   |   TOK_TYPE_ID "node" (7.26.16) typeid
|   |   |   |   TOK_IDENT "tmp" (7.26.22) struct variable (7.26.26)
|   |   |   |   TOK_ALLOC "alloc" (7.26.28)
|   |   |   |   |   TOK_TYPE_ID "node" (7.26.41) typeid
|   |   |   TOK_BLOCK "{" (7.27.4)
|   |   |   |   TOK_IF "if" (7.27.5)
|   |   |   |   |   TOK_NOT "not" (7.27.9)
|   |   |   |   |   |   TOK_NE "!=" (7.27.20)
|   |   |   |   |   |   |   TOK_IDENT "stack" (7.27.14) variable (7.14.12)
|   |   |   |   |   |   |   TOK_NULLPTR "nullptr" (7.27.23) null const
|   |   |   |   |   TOK_CALL "(" (7.27.38)
|   |   |   |   |   |   TOK_IDENT "fail" (7.27.33) variable
|   |   |   |   |   |   TOK_STRINGCON ""stack != nullptr"" (7.27.39) string const
|   |   |   |   |   |   |   TOK_STRINGCON ""41-linkedstack.oc"" (7.27.59) string const
|   |   |   |   |   |   |   TOK_INTCON "27" (7.27.80) int const
|   |   |   ';' ";" (7.27.85)
|   |   |   '=' "=" (7.28.14) int vreg
|   |   |   |   TOK_ARROW "->" (7.28.7) lval vaddr
|   |   |   |   |   TOK_IDENT "tmp" (7.28.4) struct variable (7.26.26)
|   |   |   |   |   TOK_IDENT "data" (7.28.9) variable
|   |   |   |   TOK_IDENT "str" (7.28.16) string variable (6.26.14)
|   |   |   '=' "=" (7.29.14) int vreg
|   |   |   |   TOK_ARROW "->" (7.29.7) lval vaddr
|   |   |   |   |   TOK_IDENT "tmp" (7.29.4) struct variable (7.26.26)
|   |   |   |   |   TOK_IDENT "link" (7.29.9) variable
|   |   |   |   TOK_ARROW "->" (7.29.21) lval vaddr
|   |   |   |   |   TOK_IDENT "stack" (7.29.16) variable (7.14.12)
|   |   |   |   |   TOK_IDENT "top" (7.29.23) variable
|   |   |   '=' "=" (7.30.15) int vreg
|   |   |   |   TOK_ARROW "->" (7.30.9) lval vaddr
|   |   |   |   |   TOK_IDENT "stack" (7.30.4) variable (7.14.12)
|   |   |   |   |   TOK_IDENT "top" (7.30.11) variable
|   |   |   |   TOK_IDENT "tmp" (7.30.17) struct variable (7.26.26)
|   TOK_FUNCTION "" (7.33.1) function
|   |   TOK_TYPE_ID "" (7.33.1) typeid
|   |   |   TOK_STRING "string" (7.33.1) string
|   |   |   TOK_IDENT "pop" (7.33.8) variable
|   |   TOK_PARAM "(" (7.33.12) param
|   |   |   TOK_TYPE_ID "" (7.33.13) typeid
|   |   |   |   TOK_PTR "ptr" (7.33.13)
|   |   |   |   |   TOK_TYPE_ID "stack" (7.33.24) typeid
|   |   |   |   TOK_IDENT "stack" (7.33.31) variable (7.14.12)
|   |   TOK_BLOCK "{" (7.33.38)
|   |   |   TOK_VARDECL "=" (7.34.15) variable lval
|   |   |   |   TOK_STRING "string" (7.34.4) string
|   |   |   |   TOK_IDENT "tmp" (7.34.11) struct variable (7.26.26)
|   |   |   |   TOK_ARROW "->" (7.34.27) lval vaddr
|   |   |   |   |   TOK_ARROW "->" (7.34.22) lval vaddr
|   |   |   |   |   |   TOK_IDENT "stack" (7.34.17) variable (7.14.12)
|   |   |   |   |   |   TOK_IDENT "top" (7.34.24) variable
|   |   |   |   |   TOK_IDENT "data" (7.34.29) variable
|   |   |   TOK_BLOCK "{" (7.35.4)
|   |   |   |   TOK_IF "if" (7.35.5)
|   |   |   |   |   TOK_NOT "not" (7.35.9)
|   |   |   |   |   |   TOK_NE "!=" (7.35.20)
|   |   |   |   |   |   |   TOK_IDENT "stack" (7.35.14) variable (7.14.12)
|   |   |   |   |   |   |   TOK_NULLPTR "nullptr" (7.35.23) null const
|   |   |   |   |   TOK_CALL "(" (7.35.38)
|   |   |   |   |   |   TOK_IDENT "fail" (7.35.33) variable
|   |   |   |   |   |   TOK_STRINGCON ""stack != nullptr"" (7.35.39) string const
|   |   |   |   |   |   |   TOK_STRINGCON ""41-linkedstack.oc"" (7.35.59) string const
|   |   |   |   |   |   |   TOK_INTCON "35" (7.35.80) int const
|   |   |   ';' ";" (7.35.85)
|   |   |   TOK_BLOCK "{" (7.36.4)
|   |   |   |   TOK_IF "if" (7.36.5)
|   |   |   |   |   TOK_NOT "not" (7.36.9)
|   |   |   |   |   |   TOK_NOT "not" (7.36.14)
|   |   |   |   |   |   |   TOK_CALL "(" (7.36.24)
|   |   |   |   |   |   |   |   TOK_IDENT "empty" (7.36.18) variable
|   |   |   |   |   |   |   |   TOK_IDENT "stack" (7.36.25) variable (7.14.12)
|   |   |   |   |   TOK_CALL "(" (7.36.39)
|   |   |   |   |   |   TOK_IDENT "fail" (7.36.34) variable
|   |   |   |   |   |   TOK_STRINGCON ""not empty (stack)"" (7.36.40) string const
|   |   |   |   |   |   |   TOK_STRINGCON ""41-linkedstack.oc"" (7.36.61) string const
|   |   |   |   |   |   |   TOK_INTCON "36" (7.36.82) int const
|   |   |   ';' ";" (7.36.87)
|   |   |   '=' "=" (7.37.15) int vreg
|   |   |   |   TOK_ARROW "->" (7.37.9) lval vaddr
|   |   |   |   |   TOK_IDENT "stack" (7.37.4) variable (7.14.12)
|   |   |   |   |   TOK_IDENT "top" (7.37.11) variable
|   |   |   |   TOK_ARROW "->" (7.37.27) lval vaddr
|   |   |   |   |   TOK_ARROW "->" (7.37.22) lval vaddr
|   |   |   |   |   |   TOK_IDENT "stack" (7.37.17) variable (7.14.12)
|   |   |   |   |   |   TOK_IDENT "top" (7.37.24) variable
|   |   |   |   |   TOK_IDENT "link" (7.37.29) variable
|   |   |   TOK_RETURN "return" (7.38.4)
|   |   |   |   TOK_IDENT "tmp" (7.38.11) struct variable (7.26.26)
|   TOK_FUNCTION "" (7.41.1) function
|   |   TOK_TYPE_ID "" (7.41.1) typeid
|   |   |   TOK_INT "int" (7.41.1) int
|   |   |   TOK_IDENT "main" (7.41.5) variable
|   |   TOK_PARAM "(" (7.41.10) param
|   |   |   TOK_TYPE_ID "" (7.41.11) typeid
|   |   |   |   TOK_INT "int" (7.41.11) int
|   |   |   |   TOK_IDENT "argc" (7.41.15) int variable (7.41.11)
|   |   |   TOK_TYPE_ID "" (7.41.21) typeid
|   |   |   |   TOK_ARRAY "array" (7.41.21) array
|   |   |   |   |   TOK_TYPE_ID "string" (7.41.27) typeid
|   |   |   |   TOK_IDENT "argv" (7.41.35) array variable (7.41.21)
|   |   TOK_BLOCK "{" (7.41.41)
|   |   |   TOK_VARDECL "=" (7.42.13) variable lval
|   |   |   |   TOK_INT "int" (7.42.4) int
|   |   |   |   TOK_IDENT "argi" (7.42.8) int variable (7.42.13)
|   |   |   |   TOK_INTCON "0" (7.42.15) int const
|   |   |   TOK_VARDECL "=" (7.43.28) variable lval
|   |   |   |   TOK_PTR "ptr" (7.43.4)
|   |   |   |   |   TOK_TYPE_ID "stack" (7.43.15) typeid
|   |   |   |   TOK_IDENT "stack" (7.43.22) variable (7.14.12)
|   |   |   |   TOK_CALL "(" (7.43.39)
|   |   |   |   |   TOK_IDENT "new_stack" (7.43.30) variable
|   |   |   TOK_WHILE "while" (7.44.4)
|   |   |   |   '<' "<" (7.44.16)
|   |   |   |   |   TOK_IDENT "argi" (7.44.11) int variable (7.42.13)
|   |   |   |   |   TOK_IDENT "argc" (7.44.18) int variable (7.41.11)
|   |   |   |   TOK_BLOCK "{" (7.44.24)
|   |   |   |   |   TOK_CALL "(" (7.45.12)
|   |   |   |   |   |   TOK_IDENT "push" (7.45.7) variable
|   |   |   |   |   |   TOK_IDENT "stack" (7.45.13) variable (7.14.12)
|   |   |   |   |   |   |   TOK_INDEX "[" (7.45.24) lval vaddr
|   |   |   |   |   |   |   |   TOK_IDENT "argv" (7.45.20) array variable (7.41.21)
|   |   |   |   |   |   |   |   TOK_IDENT "argi" (7.45.25) int variable (7.42.13)
|   |   |   |   |   '=' "=" (7.46.12) int vreg
|   |   |   |   |   |   TOK_IDENT "argi" (7.46.7) int variable (7.42.13)
|   |   |   |   |   |   '+' "+" (7.46.19) int vreg
|   |   |   |   |   |   |   TOK_IDENT "argi" (7.46.14) int variable (7.42.13)
|   |   |   |   |   |   |   TOK_INTCON "1" (7.46.21) int const
|   |   |   TOK_WHILE "while" (7.48.4)
|   |   |   |   TOK_NOT "not" (7.48.11)
|   |   |   |   |   TOK_CALL "(" (7.48.21)
|   |   |   |   |   |   TOK_IDENT "empty" (7.48.15) variable
|   |   |   |   |   |   TOK_IDENT "stack" (7.48.22) variable (7.14.12)
|   |   |   |   TOK_BLOCK "{" (7.48.30)
|   |   |   |   |   TOK_CALL "(" (7.49.14)
|   |   |   |   |   |   TOK_IDENT "putstr" (7.49.7) variable
|   |   |   |   |   |   TOK_CALL "(" (7.49.19)
|   |   |   |   |   |   |   TOK_IDENT "pop" (7.49.15) variable
|   |   |   |   |   |   |   TOK_IDENT "stack" (7.49.20) variable (7.14.12)
|   |   |   |   |   TOK_CALL "(" (7.50.14)
|   |   |   |   |   |   TOK_IDENT "putchr" (7.50.7) variable
|   |   |   |   |   |   TOK_CHARCON "'\n'" (7.50.15) int const
|   |   |   TOK_RETURN "return" (7.52.4)
|   |   |   |   TOK_INTCON "0" (7.52.11) int const
