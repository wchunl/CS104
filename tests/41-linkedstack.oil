fail:     .function
          .param ptr expr
          .param ptr file
          .param int line
          .end
putchr:   .function
          .param int chr
          .end
putint:   .function
          .param int num
          .end
putstr:   .function
          .param ptr str
          .end
getchr:   .function int
          .end
getstr:   .function
          .end
getln:    .function
          .end
          .struct node
          .field ptr data
          .field ptr node
          .end
          .struct stack
          .field ptr node
          .end
empty:    .function int
          .param ptr stack
.if0:     $t0:i = stack != nullptr
          $t1:i = not $t0:i
          goto .el0 if not $t1:i
.th0:     $t2:i = "stack != nullptr"
          fail($t2:i)
.fi0:     return ==
          return 
          .end
new_stack:.function
          .local ptr stack
          malloc (size of struct stack)
          stack = $t3:i
          stack->stack.top = nullptr
          return stack
          return 
          .end
push:     .function
          .param ptr stack
          .param ptr str
          .local ptr tmp
          malloc (size of struct node)
          tmp = $t4:i
.if1:     $t5:i = stack != nullptr
          $t6:i = not $t5:i
          goto .el1 if not $t6:i
.th1:     $t7:i = "stack != nullptr"
          fail($t7:i)
.fi1:     tmp->node.data = str
          $t8:p = stack->stack.top
          tmp->node.link = $t8:i
          stack->stack.top = tmp
          .end
pop:      .function
          .param ptr stack
          .local ptr tmp
          $t9:p = ->->null.data
          tmp = $t9:i
.if2:     $t10:i = stack != nullptr
          $t11:i = not $t10:i
          goto .el2 if not $t11:i
.th2:     $t12:i = "stack != nullptr"
          fail($t12:i)
.fi2:     
.if3:     empty(stack)
          $t14:i = not $t13:i
          $t15:i = not $t14:i
          goto .el3 if not $t15:i
.th3:     $t16:i = "not empty (stack)"
          fail($t16:i)
.fi3:     $t17:p = ->->null.link
          stack->stack.top = $t17:i
          return tmp
          return 
          .end
main:     .function int
          .param int argc
          .param ptr argv
          .local int argi
          .local ptr stack
          argi = 0
          new_stack()
          stack = $t18:i
.wh0:     $t19:i = argi < argc
          goto .od0 if not $t19:i
.do0:     $t20:i = stack
          push($t20:i)
          $t21:i = argi + 1
          argi = $t21:i
.od0:     
.wh1:     empty(stack)
          $t23:i = not $t22:i
          goto .od1 if not $t23:i
.do1:     pop(stack)
          putstr($t24:i)
          putchr('\n')
.od1:     return 0
          return 
          .end
