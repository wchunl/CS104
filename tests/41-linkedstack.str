string_set[   2]:    8689104156805002394 0x1c66548->""not empty (stack)""
                    5320480286546035437 0x1c61e08->"stack"
                   12597418120160561823 0x1c5fd18->"int"
string_set[   5]:    1370730858159036685 0x1c67fb8->"+"
string_set[   8]:   13497578857207247412 0x1c5fbb8->"string"
string_set[   9]:   10319632053171904911 0x1c67548->"while"
string_set[  10]:    9201800968376702500 0x1c60fa8->"getchr"
string_set[  11]:   14805121035671447500 0x1c62ec8->"->"
string_set[  12]:   14643508047410943861 0x1c671d8->"argv"
                   14752253762927134915 0x1c61b38->"data"
string_set[  14]:   15816506932648033945 0x1c5b328->"("
                    1396094175042499165 0x1c61c78->"<"
string_set[  15]:   12868445110721718657 0x1c5ff18->")"
string_set[  17]:    6142509188972423790 0x1c5b6d8->""
string_set[  20]:    1583576014053361350 0x1c60b78->"str"
string_set[  21]:     769468856621284145 0x1c64b68->"27"
                    9317282556992397945 0x1c62668->"if"
string_set[  22]:   10159970873491820195 0x1c68068->"1"
string_set[  23]:     232592375681467293 0x1c64498->"tmp"
string_set[  24]:     370180285156928338 0x1c62198->"empty"
string_set[  26]:   12291392672316406109 0x1c600d8->"line"
                    9611428142580219387 0x1c5b6f8->"fail"
                   11597697714117577063 0x1c63658->"="
string_set[  28]:   18413295556186398326 0x1c602e8->"putint"
string_set[  29]:     549611050806166819 0x1c65348->"pop"
                   11911796782241914524 0x1c5f7d8->"void"
                    2699919699724185070 0x1c61e58->"link"
string_set[  33]:    6814028903552918164 0x1c67048->"array"
string_set[  35]:   11868120863286193964 0x1c66c88->"main"
string_set[  39]:   18345866965089363306 0x1c618a8->"node"
string_set[  40]:   13945322250941355849 0x1c61718->"struct"
string_set[  41]:   12051031859636400188 0x1c67e28->"]"
string_set[  42]:   13686721134187162920 0x1c61c28->"}"
string_set[  43]:   14849783028996024063 0x1c62778->"not"
string_set[  49]:     968478856765378015 0x1c666e8->"36"
                   12679825168387489523 0x1c609c8->"num"
                    1091234871973478852 0x1c62ca8->""41-linkedstack.oc""
string_set[  50]:      21606623943142546 0x1c60288->"{"
string_set[  52]:    5344662657899890615 0x1c600f8->";"
string_set[  54]:   16250393894591675649 0x1c66068->"35"
                    7296690267450279874 0x1c605a8->"chr"
                    2497302301362042119 0x1c62e78->"new_stack"
string_set[  55]:    2297668033614959926 0x1c67478->"0"
                   17176659130743002837 0x1c61c08->"ptr"
string_set[  56]:    3701972582333163920 0x1c66c48->"argc"
string_set[  59]:    6192799629375248351 0x1c68708->"'\n'"
string_set[  60]:    1815486778388015849 0x1c63b78->"push"
string_set[  63]:   14526079891305879776 0x1c61de8->">"
string_set[  64]:   16579916199511856337 0x1c614a8->"getln"
string_set[  67]:    4706206694200543916 0x1c5fde8->","
                   12808232827189627082 0x1c62998->"nullptr"
string_set[  69]:    5266873771696973564 0x1c66f18->"argi"
string_set[  70]:     603146580336197870 0x1c62d48->"15"
string_set[  73]:    9386170080689500517 0x1c5fc98->"expr"
string_set[  75]:   16247530808313002034 0x1c621b8->"top"
string_set[  76]:   11166701219533395393 0x1c61218->"getstr"
string_set[  80]:    8290819294854170237 0x1c636a8->"alloc"
                    7211130878751977441 0x1c60438->"putchr"
                    2617056156435215158 0x1c628e8->"!="
string_set[  83]:   13581035385482130682 0x1c677f8->"["
string_set[  85]:    4017852018407011111 0x1c5fef8->"file"
string_set[  87]:   14635384904416252381 0x1c62e58->"return"
string_set[  89]:   13348923631108115385 0x1c60758->"putstr"
string_set[  92]:   11201336617819511013 0x1c62b88->""stack != nullptr""
string_set[  95]:   14736352883882276958 0x1c62f78->"=="
load_factor = 0.670
bucket_count = 97
max_bucket_size = 3
