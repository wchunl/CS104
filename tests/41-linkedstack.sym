fail (6.22.1) {0} void function
   expr (6.22.12) {1} string variable param lval 0
   file (6.22.25) {1} string variable param lval 1
   line (6.22.38) {1} int variable param lval 2

putchr (6.24.1) {1} void function
   chr (6.24.14) {1} int variable param lval 0

putint (6.25.1) {2} void function
   num (6.25.14) {1} int variable param lval 0

putstr (6.26.1) {3} void function
   str (6.26.14) {1} string variable param lval 0

getchr (6.28.1) {4} int function

getstr (6.29.1) {5} string function

getln (6.30.1) {6} string function

node (7.5.1) {0} struct node
   data (7.6.4)  string field 0
   node (7.7.4) ptr <struct node> field 1

stack (7.10.1) {0} struct stack
   node (7.11.4) ptr <struct node> field 0

empty (7.14.1) {7} int function
   stack (7.14.12) {1} ptr <struct stack> variable param lval 0

new_stack (7.19.1) {8} ptr <struct stack> function
   stack (7.20.28) {1} ptr <struct stack> variable local lval 0

push (7.25.1) {9} void function
   stack (7.25.12) {1} ptr <struct stack> variable param lval 0
   str (7.25.37) {1} string variable param lval 1
   tmp (7.26.26) {1} ptr <struct node> variable local lval 0

pop (7.33.1) {10} string function
   stack (7.33.13) {1} ptr <struct stack> variable param lval 0
   tmp (7.34.15) {1} string variable local lval 0

main (7.41.1) {11} int function
   argc (7.41.11) {1} int variable param lval 0
   argv (7.41.21) {1} array <string> variable param lval 1
   argi (7.42.13) {1} int variable local lval 0
   stack (7.43.28) {1} ptr <struct stack> variable local lval 1
