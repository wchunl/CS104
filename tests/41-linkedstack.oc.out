; 0.  1: # 1 "41-linkedstack.oc"; 1.  0: 
; 1.  1: # 1 "<built-in>"
; 2.  1: # 1 "<command-line>"
; 3.  1: # 1 "41-linkedstack.oc"
; 4.  1: 
; 4.  2: 
; 4.  3: # 1 "oclib.h" 1
; 5.  1: # 22 "oclib.h"
; 6. 22: void fail (string expr, string file, int line);
; 6. 23: 
; 6. 24: void putchr (int chr);
; 6. 25: void putint (int num);
; 6. 26: void putstr (string str);
; 6. 27: 
; 6. 28: int getchr();
; 6. 29: string getstr();
; 6. 30: string getln();
; 6. 31: # 4 "41-linkedstack.oc" 2
; 7.  4: 
; 7.  5: struct node {
; 7.  6:    string data;
; 7.  7:    ptr <struct node> link;
; 7.  8: };
; 7.  9: 
; 7. 10: struct stack {
; 7. 11:    ptr <struct node> top;
; 7. 12: };
; 7. 13: 
; 7. 14: int empty (ptr<struct stack> stack) {
; 7. 15:    {if (not (stack != nullptr)) fail ("stack != nullptr", "41-linkedstack.oc", 15);};
; 7. 16:    return stack->top == nullptr;
; 7. 17: }
; 7. 18: 
; 7. 19: ptr<struct stack> new_stack() {
; 7. 20:    ptr<struct stack> stack = alloc<struct stack>();
; 7. 21:    stack->top = nullptr;
; 7. 22:    return stack;
; 7. 23: }
; 7. 24: 
; 7. 25: void push (ptr<struct stack> stack, string str) {
; 7. 26:    ptr <struct node> tmp = alloc<struct node>();
; 7. 27:    {if (not (stack != nullptr)) fail ("stack != nullptr", "41-linkedstack.oc", 27);};
; 7. 28:    tmp->data = str;
; 7. 29:    tmp->link = stack->top;
; 7. 30:    stack->top = tmp;
; 7. 31: }
; 7. 32: 
; 7. 33: string pop (ptr<struct stack> stack) {
; 7. 34:    string tmp = stack->top->data;
; 7. 35:    {if (not (stack != nullptr)) fail ("stack != nullptr", "41-linkedstack.oc", 35);};
; 7. 36:    {if (not (not empty (stack))) fail ("not empty (stack)", "41-linkedstack.oc", 36);};
; 7. 37:    stack->top = stack->top->link;
; 7. 38:    return tmp;
; 7. 39: }
; 7. 40: 
; 7. 41: int main (int argc, array<string> argv) {
; 7. 42:    int argi = 0;
; 7. 43:    ptr<struct stack> stack = new_stack();
; 7. 44:    while (argi < argc) {
; 7. 45:       push (stack, argv[argi]);
; 7. 46:       argi = argi + 1;
; 7. 47:    }
; 7. 48:    while (not empty (stack)) {
; 7. 49:       putstr (pop (stack));
; 7. 50:       putchr ('\n');
; 7. 51:    }
; 7. 52:    return 0;
; 7. 53: }
